src/Passes/Source/Logging/ILogger.hpp

```diff
@@ -0,0 +1,58 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Types/Types.hpp"
+
+#include <cstdint>
+#include <string>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    /// Logger interface to allow the collection of different types of messages during QIR
+    /// transformation and/or validation.
+    class ILogger
+    {
+      public:
+        // Constructors, copy and move operators and destructors
+        //
+
+        ILogger()               = default;
+        ILogger(ILogger const&) = default;
+        ILogger(ILogger&&)      = default;
+        ILogger& operator=(ILogger const&) = default;
+        ILogger& operator=(ILogger&&) = default;
+
+        virtual ~ILogger() = default;
+
+        // Abstract interface methods
+        //
+
+        /// Reports a debug message.
+        virtual void debug(String const& message) = 0;
+
+        /// Reports an info message.
+        virtual void info(String const& message) = 0;
+
+        /// Reports a warning message.
+        virtual void warning(String const& message) = 0;
+
+        /// Reports an error message.
+        virtual void error(String const& message) = 0;
+
+        /// Reports an internal error message.
+        virtual void internalError(String const& message) = 0;
+
+        /// Sets the current location. Importantly, the location can be set independently of the reported
+        /// messages. This allows one to update the location upon updating the cursor position without
+        /// having to worry about keeping a copy of the location to pass when reporting messages.
+        virtual void setLocation(String const& name, uint64_t row, uint64_t col) = 0;
```

Comment:
Looking just at this interface `ILogger` (without diving into the implementation in the derived classes) I get the following questions:

> Sets the current **location**.

Location of what? Of output? If yes then may be comment that and rename to `setOutputLocation()`? 

Why does the setting of the location have the `name` parameter? How does "name" relate to the location?

Reply:
It sets the location (or cursor if you like) of the source. As a tool traverse the code it can set the location of the message being logged. This may include a name and a x, y position. The most obvious case of this is filename (name) with a line and character (row, col).

Reply:
Consider adding to the code comments something like:
> The most obvious case of this is file**path** (name) with a line and character (row, col).

Reply:
Fixed.